plugin/pulpcore/plugin/download/asyncio/group.py

```diff
@@ -0,0 +1,190 @@
+            if content_unit.done:
+                self.groups_not_done.pop(index)
+                return content_unit
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        """
+        Returns:
+            tuple: The first element is the `id` first specified by the user. The second element is
+                a dictionary where the keys are :class:`pulpcore.plugin.models.DeferredArtifact` and
+                the values are dictionaries containing information about the downloaded file. The
+                keys of this dictionary are 'filename', 'size', 'sha1' , 'md5', 'sha224',
+                'sha256', 'sha384', and 'sha512'.
+        """
+        while self.downloads_not_done:
+            done_this_time, self.downloads_not_done = \
+                self.loop.run_until_complete(asyncio.wait(self.downloads_not_done,
+                                                          return_when=asyncio.FIRST_COMPLETED))
+            with suppress(StopIteration):
+                if self.group_iterator:
+                    self.schedule_group(**next(self.group_iterator))
+            for task in done_this_time:
+                download_result = task.result()
+                for group in self.urls[download_result.url]:
+                    group.handle_download_result(download_result)
+                group = self._find_and_remove_done_group()
+                if group:
+                    return group.finished_downloads
+        content_unit = self._find_and_remove_done_group()
+        if content_unit:
+            return content_unit.finished_downloads
+        else:
+            raise StopIteration()
+
+
+class Group(object):
+
+    def __init__(self, id, deferred_artifacts):
+        """
+        Args:
+            id (hashable): This id is handed back to the caller when all files in the group are
+                downloaded.
+            deferred_artifacts (list): A list of :class:`pulpcore.plugin.models.DeferredArtifact`
+                instances that have not been saved to the database.
+        """
+        self.id = id
+        self.deferred_artifacts = {}
+        self.downloaded_files = {}
+        urls = []
+        for deferred_artifact in deferred_artifacts:
+            self.deferred_artifacts[deferred_artifact.url] = deferred_artifact
+            urls.append(deferred_artifact.url)
+        self.urls = set(urls)
+        self.finished_urls = []
+
+    def handle_download_result(self, download_result):
+        """
+        Update the Group with download result calculated during the download from the URL
+
+        Args:
+            download_result (DownloadResult): The return argument from an HttpDownloader
+        """
+        self.finished_urls.append(download_result.url)
+        self.downloaded_files[download_result.url] = download_result
+
+    @property
+    def done(self):
+        return len(self.urls) == len(self.finished_urls)
+
+    @property
+    def finished_downloads(self):
```

Comment:
Please add docstring with `Returns:`.

Reply:
I *think* it returns `(id, DownloadResult)`.  Right?

Reply:
Yes that is what it returns. This is added in the next push along with one for the `done` property too.